Adds design template

irbekrm opened the pull request:
This PR adds a design template as discussed earlier in dev meetings.

This template could be used by folks who want to contribute a more complex feature/enhancement to cert-manager and need to submit a design beforehand. The general approach of requiring a design should help us improve the review process and make it easier for maintainers and users to understand and evaluate new features.



The template partially adapted from https://github.com/kubernetes/enhancements/blob/a86942e8ba802d0035ec7d4a9c992f03bca7dce9/keps/NNNN-kep-template/README.md - I have tried to remove the more strict or unusual for us requirements as generally I would like to _not_ introduce new 'policies' in this PR- they can always be added later separately as need arises.

If this gets merged, I will update our contributing doc to point to this https://cert-manager.io/docs/contributing/policy/#design-documents

Happy to remove more sections from the template if folks think it's too complicated.
Generally found it very useful to read through the upstream template- there are a lot of questions that relate to actual risks when adding new features to cert-manager (i.e increase in memory usage, extra API calls) so I think having this will be very useful for reviewers.

Signed-off-by: irbekrm <irbekrm@gmail.com>

irbekrm commented:
/kind cleanup

irbekrm commented on design/template.md:
```diff
@@ -0,0 +1,211 @@
+For example, are external dependencies such as ingress controllers, third party CRDs etc required for this feature to function?
+-->
+
+###### Will enabling / using this feature result in any new API calls (to Kubernetes or external services)?
```
This will be good to consider as it will include extra calls to external services like LetsEncrypt

irbekrm commented on design/template.md:
```diff
@@ -0,0 +1,211 @@
+
+###### Will enabling / using this feature result in introducing new API types?
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
```
i.e whether more CertificateRequests will be created

SgtCoDFish replied:
This might be worth adding in a comment just to make it clearer which API objects we're talking about!

irbekrm replied:
It could be core types as well, i.e more or larger `Secret`s
But good point, will add a comment

irbekrm commented on design/template.md:
```diff
@@ -0,0 +1,211 @@
+
+
+
+### Graduation Criteria
```
I have tried to keep this relatively non-specific as I think we can agree on more strict graduation criteria later if needed- perhaps this needs to made even less specific

irbekrm commented:
/test pull-cert-manager-make-test

irbekrm commented:
/test pull-cert-manager-e2e-v1-23

irbekrm commented:
/test pull-cert-manager-e2e-v1-23
/test pull-cert-manager-make-e2e-v1-23

irbekrm commented:
/test pull-cert-manager-e2e-v1-23
/test pull-cert-manager-make-e2e-v1-23

irbekrm commented:
/test pull-cert-manager-make-e2e-v1-23

SgtCoDFish commented on design/template.md:
```diff
@@ -0,0 +1,232 @@
+This section is important for producing high-quality, user-focused
+documentation such as release notes.
+
+A good summary is probably at least a paragraph in length.
```
suggestion (non-blocking): I think I'd probably want a summary to usually be _around_ a paragraph in length, or even _exactly_ one paragraph in most cases! I don't think that's important though - I think people will naturally tend towards shorter summaries.

SgtCoDFish commented on design/template.md:
```diff
@@ -0,0 +1,232 @@
+This template is adapted from Kubernetes Enchancements KEP template https://raw.githubusercontent.com/kubernetes/enhancements/a86942e8ba802d0035ec7d4a9c992f03bca7dce9/keps/NNNN-kep-template/README.md
```
suggestion: By having this be actual text I think there's a decent chance that people won't remove it and it'll get in the way of the actual design doc.

I think it's nice to have this link but it could be a comment; failing that maybe we could move it right to the bottom of the template?

```suggestion
<!--
This template is adapted from Kubernetes Enchancements KEP template https://raw.githubusercontent.com/kubernetes/enhancements/a86942e8ba802d0035ec7d4a9c992f03bca7dce9/keps/NNNN-kep-template/README.md
-->
```

irbekrm replied:
Comment make sense yeah 👍🏼

SgtCoDFish commented on design/template.md:
```diff
@@ -0,0 +1,232 @@
+This template is adapted from Kubernetes Enchancements KEP template https://raw.githubusercontent.com/kubernetes/enhancements/a86942e8ba802d0035ec7d4a9c992f03bca7dce9/keps/NNNN-kep-template/README.md
+
+# Title
```
suggestion: I think this is the only title we'd _always_ want people to change; maybe we can make it clearer that the user should change it?

```suggestion
# CHANGEME: Title
```

SgtCoDFish commented on design/template.md:
```diff
@@ -0,0 +1,232 @@
+
+## Release Signoff Checklist
+
+These checklist contains actions that must be completed before the PR with the feature described in this design can be merged.
```
suggestion: typo at beginning, change wording slightly

```suggestion
This checklist contains actions which must be completed before a PR implementing this design can be merged.
```

SgtCoDFish commented on design/template.md:
```diff
@@ -0,0 +1,232 @@
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go into as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
```
suggestion (non-blocking): I think this section could maybe be moved down to the bottom since it's optional?

irbekrm replied:
I'd rather keep this here as I think when reading from top to bottom it makes sense to have these additional details before the more implementation specific stuff (tests, API changes) - to give the reader the ability to get the high level overview of the proposal

irbekrm replied:
I do imagine that it would be unused in most cases though

SgtCoDFish commented on design/template.md:
```diff
@@ -0,0 +1,232 @@
+What versions of Kubernetes (and other external services if applicable) will this feature support?
+-->
+
+## Production Readiness
+<!--
+This section should confirm that the feature can be safely operated in production environment and can be disabled or rolled back in case it is found to increase failures.
+-->
+
+
+###### How can this feature be enabled / disabled for an existing cert-manager installation?
+
```
suggestion (non-blocking): we jump from `##` to `######` here; why not `###`? 

(I don't feel strongly on this but I usually try to only go down one level)

SgtCoDFish reviewed (approved):
/lgtm
/hold

I put a bunch of comments, I don't really feel any of them have to be blockers really! Overall I see no reason not to merge because having this is probably better than not having this!

But I'm a little concerned that this might be a little too _heavyweight_? Like, this is quite a lot to do for someone who wants a new feature, and I feel like it could turn someone who would otherwise have contributed away.

Could there maybe be a small subset of these questions which people answer first, and then if we're concerned we ask them to answer all of the questions just to be sure?

Still, that's all nitpicking and I'd happily merge as-is so I'm approving with a hold!

irbekrm commented:
Thanks a lot for taking a look @SgtCoDFish !

I think I have addressed all the comments.

About whether it's heavyweight, it doesn't seem like that to me, but I happy to have a discussion and a vote on this if you prefer?
Perhaps we could also try to get some feedback from first users who attempt to use the template.

Some thoughts- 

I am keen on the template as there have been a couple cases where users have described an issue and their preferred solution  via the currently existing channels- creating GH issues and mentioning it on Slack, but gotten no response- perhaps due to our low bandwidth and the fact that the solution that may seem obvious to them is full of unknowns for us (I talked to one such potential contributor at our booth in Kubecon)- by giving them a format how to answer all our possible concerns we can partially fix the current state where the amount of initial investigation and later consideration of all possible edge cases is so much work for maintainers that feature suggestions are not considered and/or PRs don't get reviewed unless the contributor does some social engineering to get it noticed :) 

> Could there maybe be a small subset of these questions which people answer first, and then if we're concerned we ask them to answer all of the questions just to be sure?

Maybe, but by putting them in one place means that we ourselves will also not forget to consider these aspects- I certainly could easily forget to think about what could be the potential resource increase or impact from extra API calls- so from that perspective feels good if it's there. Again happy to have a vote on this if you think it's needed!

I don't know if people would actually be deterred from adding a feature if they need to fill in design beforehand- I do see that a lot of folks that create issues on Github and remove all the bits of template even the bits that would have actually been useful- however not sure if that's just a small minority

irbekrm commented:
/test pull-cert-manager-e2e-v1-23

SgtCoDFish reviewed (approved):
/approve
/lgtm

Thanks for looking at my comments!

> but I happy to have a discussion and a vote on this if you prefer?

I don't really think this needs to go to a vote or that it really needs more discussion. I think we should give it a go and it'll be better than what we have now! If we need to change it down the road we always can.

I'll leave the hold in place but I'm happy to merge this :+1:
